### ISSUE TITLE
Dependency Jars on hdfs with High Available NameNode unable to resolve

### ISSUE THREAD

**Issue by ravadisachin** [GoogleCloudPlatform/spark-on-k8s-operator#572]
When I keep Dependency Jars on a hdfs cluster with NameNode High Availabilty configured and dfs.nameservices set to namenodeHA,the Spark Submission fails (SUBMISSION_FAILED state).
      "deps": {
                "jars": [
                          "hdfs://namenodeHA/tmp/spark-streaming-kafka-assembly_2.11-1.6.0.jar",
                          "hdfs://namenodeHA/tmp/spark-core_2.11-1.5.2.logging.jar"
                        ]
                     },
      "hadoopConfigMap": "non-kerberos-config",

 But when I specify the namenode_host IP,the application works fine.

      "deps": {
                "jars": [
                          "hdfs://{namenode_host_IP}/tmp/spark-streaming-kafka-assembly_2.11-1.6.0.jar",
                          "hdfs://{namenode_host_IP}/tmp/spark-core_2.11-1.5.2.logging.jar"
                        ]
                     },
      "hadoopConfigMap": "non-kerberos-config",

Could you please let me know if additional configuration needs to be done to specify dfs.nameservices(namenodeHA)? 

Thanks!!

**Comment by liyinan926** [comment]
Not familiar with HDFS HA and how configuration is done. You might want to ask this question on the spark-users mailing list instead.

**Comment by abhisrao** [comment]
Looks like the issue is with respect to resolution of namenodeHA for the dependency jars. Spark-submit client (which is operator pod in this case) is not able to resolve the namenodeHA to an IP. This will be taken care if we can copy the site-xml's inside the operator pod and set HADOOP_CONF_DIR to the location where site-xml's are present. We've tried this and it seems to be working.
Spark on kubernetes also expects HADOOP_CONF_DIR to be set during spark-submit in order to resolve the Hadoop dependencies.

**Comment by dennischin** [comment]
@abhisrao @liyinan926 Should this not be supported by passing an hdfs config map in the spark operator jobSpec and leveraging the mutating webhook?

I am experiencing the same issue and I have specified an hdfs config-map with core-site.xml and hdfs-site.xml files which properly configure for hdfs ha namesevice; however my spark submit is failing as well.

**Comment by dennischin** [comment]
@liyinan926 can we get confirmation that hdfs config map mounted files (via. webhook) are available during spark-submit time so we can utilize hdfs ha nameservice urls? 

And if not, can we make that an enhancement request or treat as a bug?

**Comment by liyinan926** [comment]
ConfigMap mounting through the webhook is only available for Spark driver and executor pods. For the operator pod, you need to customize the pod spec to add the Hadoop ConfigMap.

**Comment by dennischin** [comment]
ok.  I guess I missed the "operator pod" part of my problem.  Can you point me to documentation on how to customize the operator pod spec fo add the Hadoop ConfigMap or provide a sample?

**Comment by liyinan926** [comment]
If you can install using the manifest, simply update the pod spec in one of the operator Deployment yamls, e.g., https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/manifest/spark-operator-with-webhook.yaml.

**Comment by abhisrao** [comment]
> @abhisrao @liyinan926 Should this not be supported by passing an hdfs config map in the spark operator jobSpec and leveraging the mutating webhook?
> 
> I am experiencing the same issue and I have specified an hdfs config-map with core-site.xml and hdfs-site.xml files which properly configure for hdfs ha namesevice; however my spark submit is failing as well.

hdfs configmap via webhook will only be available to driver and executor. spark-submit client would also require this. We'll try using the option which @liyinan926 has mentioned. Thanks.

**Comment by breetasinha1109** [comment]
Hi @liyinan926 @dennischin 

As suggested by @abhisrao we've tried copying hadoop site xmls present in .spc.hadoopConfigMap to operator pod and setting HADOOP_CONF_DIR during spark-submit to the location where site-xml's are present. This way, namenodeHA for the dependency jars is being resolved and also, we are not limiting spark jobs to interact with single hadoop cluster.
Also, the site-xmls are deleted from the operator pod on completion, failure or deletion of spark application.

This is the [PR](https://github.com/nokia/spark-on-k8s-operator/pull/4) and fix is available at https://github.com/nokia/spark-on-k8s-operator/tree/master

Thanks,
Breeta

**Comment by dennischin** [comment]
I resolved the issue by modifying, spark-operator-deployment.yaml in the helm chart.

I added a volume "hadoop-configmap-volume" to reference my hdfs configmap (which specifies my hdfs core-site.xml and hdfs-site.xml files)

Created an env var "HADOOP_CONF_DIR" pointing to "/etc/hadoop/conf"

And created a voluemount for "hadoop-configmap-volume to the mountPath "/etc/hadoop/conf"

I have verified this works for me.  In the future I would like to pass the configmap name via. the values.yaml.
